### ISSUE TITLE
Add example of image segmentation task

### ISSUE THREAD

**Issue by BenjaminBossan** [dnouri/skorch#160]
There should be an example (notebook) that shows how to deal with this kind of problem. Additionally, it could show how to load images and segmentation masks on the fly.

**Comment by BlanchonMarc** [comment]
Hello Benjamin, after all your help and some experiments I am able to provide you more materials and experiments that will not restrict other people to only training.

------------------------------------------------
**Solution to perform only training**

With two separated dataset extractors inherited from PyTorch Dataset

```

n_classes = 12

network = SegNet(in_channels=3, n_classes=n_classes)
network.init_encoder()

network.cuda()

net = skorch.NeuralNet(
    module=network,
    criterion=torch.nn.CrossEntropyLoss,
    train_split=None,
    use_cuda=True,
    batch_size=5,
    callbacks__train_loss__target_extractor=lambda x: x,
    callbacks__valid_loss__target_extractor=lambda x: x
)

net.fit(X=X, y=Y)
```

In this case, the training perform correctly.


---------------------------------------------
**Solution to perform grid search**

This solution also work with the fit (only training), or also even with the grid search.

```
n_classes = 12

network = SegNet(in_channels=3, n_classes=n_classes)
network.init_encoder()

network.cuda()

net = skorch.NeuralNet(
    module=network,
    criterion=torch.nn.CrossEntropyLoss,
    train_split=None,
    use_cuda=True,
    batch_size=5,
)

params = {
    'lr': [0.01, 0.02],
    'max_epochs': [5, 10]
}

# if only training
# net.fit(X=X, y=y)

image_indicators = np.hstack([np.repeat(i, len(x)) for i, x in
                                      enumerate(X)])
labels = image_indicators % n_classes
X, y = np.vstack(X), np.hstack(Y)

cv = LeavePLabelOut(labels=labels, p=1)
gs = GridSearchCV(net, params, scoring='f1', verbose=10, cv=cv, n_jobs=-1)

gs.fit(X=X, y=y)
print(gs.best_score_, gs.best_params_)
```

Will allow a grid search after some GPU parallel processing tweaks and initializations (spawn start method).

-----------------------------------

Hope this will help you and other guys that want to perform grid search on segmentation processes.

**Comment by thomasjpfan** [comment]
I am in the process of creating an image segmentation tutorial with data from https://www.kaggle.com/c/data-science-bowl-2018. The training dataset is less than 100MB compressed, which makes it ideal for a tutorial.

**Comment by BenjaminBossan** [comment]
Just out of curiosity, what architecture will you choose? Do you use some other library for this or start from scratch?

**Comment by thomasjpfan** [comment]
Given the problem is a binary segmentation problem, I am building a UNet with a VGG11 as a feature encoder. The VGG11 weights are frozen while the decoder weights are updated.

torchvision is the only other explicit dependency. Since installing torchvision also installs Pillow, Pillow is used for any image augmentation.

**Comment by BenjaminBossan** [comment]
Nice, I'm eager to see your work. There are a couple of unet implementations in pytorch out there, though they often differ from one another and the paper(s). Do you plan on testing dilated convolutions?
